### ISSUE TITLE
[Help wanted] Debugging with GDB from Visual Studio Code

### ISSUE THREAD

**Issue by HorstBaerbel** [mgba-emu/mgba#1391]
I'm trying to get mgba debugging via GDB running using Visual Studio on Ubuntu 18.04, but I'm failing. I'm using mgba 0.8-5565-9424cd54 built from git, with Qt 5.9.5 and ELF support and installed via "sudo make install". The GDB version is 8.2.  
* What works:
Run: "mgba-qt -g" and Load the .elf in the GUI.  
Run "arm-none-eabi-gdb -ex 'target remote localhost:2345' foo.elf".  
-> I can continue, step etc.  

* What does not work:
Run: "mgba-qt -g" or "mgba-qt -g foo.elf".  
Run "arm-none-eabi-gdb -ex 'target remote localhost:2345' foo.elf".  
GDB tells me:  

  > Reading symbols from foo.elf...done.
  > Remote debugging using localhost:2345
  > Ignoring packet error, continuing...
  > warning: unrecognized item "timeout" in "qSupported" response
  > Ignoring packet error, continuing...
  > Remote replied unexpectedly to 'vMustReplyEmpty': timeout

  I get a prompt, but continue tells me "The program is not being run.".

* When trying to debug from Visual Studio Code via this [launch.json](https://github.com/Microsoft/vscode-cpptools/blob/master/launch.md):
```
        {
            "name": "Debug gbatv in mgba",
            "type": "cppdbg",
            "request": "launch",
            "preLaunchTask": "build Debug",
            "program": "${workspaceFolder}/build/foo.elf",
            "debugServerPath": "/usr/bin/mgba-qt",
            "debugServerArgs": "-2 -g ${workspaceFolder}/build/foo.elf",
            "cwd": "${workspaceRoot}/build",
            "stopAtEntry": false,
            "MIMode": "gdb",
            "miDebuggerPath": "${env:DEVKITARM}/bin/arm-none-eabi-gdb",
            //"miDebuggerArgs": "-ex \"target remote localhost:2345\" ${workspaceFolder}/build/foo.elf",
            "miDebuggerServerAddress": "localhost:2345",
            "setupCommands": [
                {
                    "text": "set print pretty",
                    "ignoreFailures": false
                },
                {
                    "text": "file ${workspaceFolder}/build/foo.elf",
                    "ignoreFailures": false
                }
            ],
            //"externalConsole": false,
            //"showDisplayString": true,
            "logging": {
                "traceResponse": true,
                "moduleLoad": true,
                "trace": true,
                "engineLogging": true,
                "programOutput": true,
                "exceptions": true
            },
        },
```
mgba is started, the elf is run and the gdb server is active, but after 10s VSCode tells me: "Unable to start debugging. Debug server process failed to initialize". I've tried all possible variations of commands in launch.json already...
This is the VSCode log of the debug process starting:

> --> E (output): {"event":"output","body":{"category":"console","output":"1: (141) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'\n"},"seq":2,"type":"event"}
> 1: (141) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  ExePath='/PATH_EDITED/build/foo.elf'\n"},"seq":4,"type":"event"}
> 1: (152) LaunchOptions  ExePath='/PATH_EDITED/build/foo.elf'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  WorkingDirectory='/PATH_EDITED/build'\n"},"seq":6,"type":"event"}
> 1: (152) LaunchOptions  WorkingDirectory='/PATH_EDITED/build'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (152) LaunchOptions  ExeArguments=''\n"},"seq":8,"type":"event"}
> 1: (152) LaunchOptions  ExeArguments=''
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (153) LaunchOptions  MIMode='gdb'\n"},"seq":10,"type":"event"}
> 1: (153) LaunchOptions  MIMode='gdb'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (153) LaunchOptions  MIDebuggerPath='/opt/devkitpro/devkitARM/bin/arm-none-eabi-gdb'\n"},"seq":12,"type":"event"}
> 1: (153) LaunchOptions  MIDebuggerPath='/opt/devkitpro/devkitARM/bin/arm-none-eabi-gdb'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (154) LaunchOptions  WaitDynamicLibLoad='false'\n"},"seq":14,"type":"event"}
> 1: (154) LaunchOptions  WaitDynamicLibLoad='false'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (154) LaunchOptions  MIDebuggerServerAddress='localhost:2345'\n"},"seq":16,"type":"event"}
> 1: (154) LaunchOptions  MIDebuggerServerAddress='localhost:2345'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (154) LaunchOptions  DebugServer='/usr/bin/mgba-qt'\n"},"seq":18,"type":"event"}
> 1: (154) LaunchOptions  DebugServer='/usr/bin/mgba-qt'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (154) LaunchOptions  DebugServerArgs='-2 -g /PATH_EDITED/build/foo.elf'\n"},"seq":20,"type":"event"}
> 1: (154) LaunchOptions  DebugServerArgs='-2 -g /PATH_EDITED/build/foo.elf'
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (155) LaunchOptions>\n"},"seq":22,"type":"event"}
> 1: (155) LaunchOptions>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (155) LaunchOptions    <SetupCommands>\n"},"seq":24,"type":"event"}
> 1: (155) LaunchOptions    <SetupCommands>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>set print pretty</Command>\n"},"seq":26,"type":"event"}
> 1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>set print pretty</Command>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>target remote localhost:2345</Command>\n"},"seq":28,"type":"event"}
> 1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>target remote localhost:2345</Command>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>file /PATH_EDITED/build/foo.elf</Command>\n"},"seq":30,"type":"event"}
> 1: (156) LaunchOptions        <Command IgnoreFailures='false' Description=''>file /PATH_EDITED/build/foo.elf</Command>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (157) LaunchOptions    </SetupCommands>\n"},"seq":32,"type":"event"}
> 1: (157) LaunchOptions    </SetupCommands>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (157) LaunchOptions</LocalLaunchOptions>\n"},"seq":34,"type":"event"}
> 1: (157) LaunchOptions</LocalLaunchOptions>
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (302) Starting: \"/usr/bin/mgba-qt\" -2 -g /PATH_EDITED/build/foo.elf\n"},"seq":36,"type":"event"}
> 1: (302) Starting: "/usr/bin/mgba-qt" -2 -g /PATH_EDITED/build/foo.elf
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (402) STDERR: QFSFileEngine::open: No file name specified\n"},"seq":38,"type":"event"}
> 1: (402) STDERR: QFSFileEngine::open: No file name specified
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (554) STDERR: QFSFileEngine::open: No file name specified\n"},"seq":40,"type":"event"}
> 1: (554) STDERR: QFSFileEngine::open: No file name specified
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (641) STDERR: QOpenGLContext::swapBuffers() called with non-exposed window, behavior is undefined\n"},"seq":42,"type":"event"}
> 1: (641) STDERR: QOpenGLContext::swapBuffers() called with non-exposed window, behavior is undefined
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (688) STDERR: QWidget::repaint: It is dangerous to leave painters active on a widget outside of the PaintEvent\n"},"seq":44,"type":"event"}
> 1: (688) STDERR: QWidget::repaint: It is dangerous to leave painters active on a widget outside of the PaintEvent
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (689) STDERR: QWidget::repaint: It is dangerous to leave painters active on a widget outside of the PaintEvent\n"},"seq":46,"type":"event"}
> 1: (689) STDERR: QWidget::repaint: It is dangerous to leave painters active on a widget outside of the PaintEvent
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (10320) <-logout\n"},"seq":48,"type":"event"}
> 1: (10320) <-logout
> --> E (output): {"event":"output","body":{"category":"console","output":"1: (10345) Send Event AD7MessageEvent\n"},"seq":50,"type":"event"}
> 1: (10345) Send Event AD7MessageEvent
> --> E (output): {"event":"output","body":{"category":"telemetry","output":"VS/Diagnostics/Debugger/Launch","data":{"VS.Diagnostics.Debugger.ImplementationName":"Microsoft.MIDebugEngine","VS.Diagnostics.Debugger.EngineVersion":"14.0.60206.1","VS.Diagnostics.Debugger.HostVersion":"14.0.60206.1","VS.Diagnostics.Debugger.AdapterId":"cppdbg","VS.Diagnostics.Debugger.Launch.ErrorCode":1005,"VS.Diagnostics.Debugger.Launch.IsError":true}},"seq":52,"type":"event"}
> --> R (launch-2): {"request_seq":2,"success":false,"command":"launch","message":"Unable to start debugging. Debug server process failed to initialize.","body":{"error":{"id":1005,"format":"Unable to start debugging. Debug server process failed to initialize."}},"seq":54,"type":"response"}
> 

It would be really nice to get debugging via VSCode working. Hope someone can help or maybe there's a bug in mgba preventing this from working. Tell me if I can try something else or if you need more input...

**Comment by paeolo** [comment]
Hi,
I didn't fully read your post but it turns out I just managed to make the gdb server work with VSCode. I assume you are using DevKitPro, so you have $DEVKITARM environnement variable set and you want to use VSCode with arm-none-eabi-gdb.

All you need is the "Native Debug" extension from WebFreak. Then you can launch mgba with the "-g" option on your ELF file, don't use the GBA file of course since it contains no debug information.

Also be sure to use the -g option for the compiler and the linker so it produces debug informations. At the end of the debugging, you can close mgba with a post debug Task on VSCode.

I finally came up with the right configs files but unfortunately you need to press enter to close mgba after exiting the debug. It's very important to do so because for some reason I suspect the socket not to be properly closed on mgba so you need to restart your session.. Anyway it's annoying to do so, I asked the vscode github how to close a task in Vscode without to press enter..

I link the config files, then you can press ctrl+shift+b for build, ctrl+shift+t for run and F5 for debug.

Enjoy.
[launch.txt](https://github.com/mgba-emu/mgba/files/3123246/launch.txt)
[tasks.txt](https://github.com/mgba-emu/mgba/files/3123247/tasks.txt)
[keybindings.txt](https://github.com/mgba-emu/mgba/files/3123249/keybindings.txt)
[c_cpp_properties.txt](https://github.com/mgba-emu/mgba/files/3123250/c_cpp_properties.txt)

**Comment by paeolo** [comment]
Better I link you a zip folder with everything and a working version of AGBPrint.
[gba-dev-debug.zip](https://github.com/mgba-emu/mgba/files/3123257/gba-dev-debug.zip)
Bye.

**Comment by HorstBaerbel** [comment]
Thanks for your help.
I tried your launch / tasks files, but it's not working for me. The elf is compiled with debug info (checked with objdump). When running the task mgba starts and runs the elf. The VSCode debugger console tells me:

> Reading symbols from /PATH_EDITED/build/foo.elf...
> done.
> Ignoring packet error, continuing...
> warning: unrecognized item "timeout" in "qSupported" response
> Ignoring packet error, continuing...

After some seconds a window pops up telling me "Failed to attach: Remote replied unexpectedly to 'vMustReplyEmpty': timeout (from target-select remote localhost:2345)".
Basically the same thing as before...

Sidenote: You can add a dependency to your "debugRun" task to always build the elf before debugging. Add:
`"dependsOn": ["build Debug"],`
(or whatever your build task label is) to your task for that.

**Comment by paeolo** [comment]
I just unzipped the "gba-dev-debug.zip", make and debug and it works. I'm using gdb 8.2 same as you but mgba-qt 0.7.1. From time to time (but not so often) i get the same error as you, I really don't know why. Does the gba-dev-debug.zip works when you use mgba-qt 0.7.1 ?

**Comment by paeolo** [comment]
ty for the sidenote

**Comment by HorstBaerbel** [comment]
Can only test with the 0.8.5 git version, because 0.7.1 needs libmagickwand-6.q16-6, but only -3 is installed by default on my current Ubuntu 18.04 :/ Don't want to fup my packages.
Your zip compiles, but has the same error when running the debugger. Maybe it is mgba-related after all...

**Comment by HorstBaerbel** [comment]
What version of devkitPro / ARM do you have?

**Comment by paeolo** [comment]
deviKitARM release 52.

**Comment by HorstBaerbel** [comment]
I've updated to devkitARM r52-1 using pacman, reconfigured, recompiled AND IT WORKS! Wtf... Thanks for your help. I'll try getting the regular launch option to work now. Will comment if that works. Thanks again!

**Comment by HorstBaerbel** [comment]
Uhm. I seem to have the socket problem.

> you need to press enter to close mgba after exiting the debug. It's very important to do so because for some reason I suspect the socket not to be properly closed on mgba so you need to restart your session..

Where do I have to press enter?

**Comment by paeolo** [comment]
lol gg! When I tried to configure the whole things, I noticed that sometimes I couldn't start the GDB server using the mgba-qt UI, had an error like "couldn't launch the server". It seemed I needed to log out/log in my linux session to get it work again, that's why I think it's a socket not properly closed or something else, I really don't know.
Anyway, there is no bug on my side if after the debugging is done I stop the debug on VSCode, ie. I click on the disconnect button (the red square). And then, you want mgba to be killed (somewhat in this order). So I'm using the postDebug Task for that but unfortunately you need to press enter on VSCode to terminate the Task (the preLaunch Task that run mgba).

**Comment by paeolo** [comment]
so the order in VSCode is : start debug (F5), do your job, stop debug (red square or shift+F5) and then press enter to kill mgba.

**Comment by HorstBaerbel** [comment]
It is on half-working for me. It breaks randomly and debugging stops working. Btw. I filed and issue some time ago for [VSCode](https://github.com/Microsoft/vscode-cpptools/issues/2982).

**Comment by HorstBaerbel** [comment]
Not working again. Has somebody found a proper solution to this?

**Comment by HorstBaerbel** [comment]
I have a working solution now (Linux, building with CMake, mGBA (GUI version) installed to /usr/local/bin, replace EXECUTABLE_NAME by the name of you executable):

launch.json:
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Run Unity in mgba",
            "type": "cppdbg",
            "request": "launch",
            "preLaunchTask": "build Release",
            "program": "/usr/local/bin/mgba-qt",
            "args": [
                "-2",
                "${command:cmake.launchTargetPath}"
            ],
            "cwd": "${command:cmake.buildDirectory}"
        },
        {
            "name": "Debug Unity in mgba",
            "type": "cppdbg",
            "request": "launch",
            "preLaunchTask": "build Debug",
            "program": "${command:cmake.buildDirectory}/EXECUTABLE_NAME.elf",
            "args": [],
            "stopAtEntry": false,
            "cwd": "${command:cmake.buildDirectory}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerServerAddress": "localhost:2345",
            "linux": {
                "miDebuggerPath": "${env:DEVKITARM}/bin/arm-none-eabi-gdb",
                "setupCommands": [
                    {
                        "text": "set print pretty",
                        "ignoreFailures": false
                    },
                    {
                        "description": "mGBA",
                        "text": "shell \"/usr/local/bin/mgba-qt\" -2 -g \"${command:cmake.buildDirectory}/EXECUTABLE_NAME.elf\" &",
                        "ignoreFailures": false
                    }
                ]
            },
            "postDebugTask": "kill mGBA"
        }
    ]
}
```

task.json:
```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "build Release",
            "type": "shell",
            "command": "cmake --build '${workspaceRoot}/build' --config Release --target all -- -j 6",
            "options": {
                "cwd": "${workspaceRoot}"
            },
            "group": {
                "kind": "build",
                "isDefault": true
            }
        },
        {
            "label": "build Debug",
            "type": "shell",
            "command": "cmake --build '${workspaceRoot}/build' --config Debug --target all -- -j 6",
            "options": {
                "cwd": "${workspaceRoot}"
            },
            "group": "build",
        },
        {
            "label": "kill mGBA",
            "group": "none",
            "type": "shell",
            "command": "killall mgba-qt"
        }
    ]
}
```

This has worked 100% for me for a couple of days now. Finally!
